modules/seqexec/server/src/main/scala/seqexec/server/tcs/TcsConfigRetriever.scala

```diff
@@ -192,6 +195,75 @@ object TcsConfigRetriever {
     getStatusVal(Nested(TcsEpics.instance.sourceAWavelength).map(v => Wavelength(Angstroms(v))).value,
       "central wavelength")
 
+  private def getGemsMap: IO[Map[GemsSource, VirtualGemsTelescope]] = (for {
+    v1 <- OptionT(TcsEpics.instance.g1MapName)
+    v2 <- OptionT(TcsEpics.instance.g2MapName)
+    v3 <- OptionT(TcsEpics.instance.g3MapName)
+    v4 <- OptionT(TcsEpics.instance.g4MapName)
+  } yield Map(
+    v1 -> VirtualGemsTelescope.G1,
+    v2 -> VirtualGemsTelescope.G2,
+    v3 -> VirtualGemsTelescope.G3,
+    v4 -> VirtualGemsTelescope.G4
+  )).value.map(_.getOrElse(Map.empty))
+
+  private def getNgs1(g: VirtualGemsTelescope, active: IO[Option[Boolean]]): IO[GuiderConfig] = for {
+    trk <- getStatusVal(getNodChopTrackingConfig(TcsEpics.instance.gemsGuideConfig(g)), "NGS1 tracking configuration")
+    fol <- getStatusVal(Nested(TcsEpics.instance.ngs1Follow)
+      .map{ if(_) FollowOption.FollowOn else FollowOption.FollowOff}.value, "NGS1 follow state")
+    wfs <- getStatusVal(Nested(active).map{if(_) GuiderSensorOn else GuiderSensorOff}.value, "NGS1 Active status")
+  } yield GuiderConfig(calcProbeTrackingConfig(fol, trk), wfs)
+
+  private def getNgs2(g: VirtualGemsTelescope, active: IO[Option[Boolean]]): IO[GuiderConfig] = for {
+    trk <- getStatusVal(getNodChopTrackingConfig(TcsEpics.instance.gemsGuideConfig(g)), "NGS2 tracking configuration")
+    fol <- getStatusVal(Nested(TcsEpics.instance.ngs2Follow)
+      .map{ if(_) FollowOption.FollowOn else FollowOption.FollowOff}.value, "NGS2 follow state")
+    wfs <- getStatusVal(Nested(active).map{if(_) GuiderSensorOn else GuiderSensorOff}.value, "NGS2 Active status")
+  } yield GuiderConfig(calcProbeTrackingConfig(fol, trk), wfs)
```

Comment:
Looks like a lot of copy/paste in this file. Is there anything you can abstract out?